### ISSUE TITLE
Bug: [prefer-optional-chain] doesn't report at all when `requireNullish` is `true`

### ISSUE THREAD

**Issue by auvred** [typescript-eslint/typescript-eslint#8487]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPJjAJYoDm8APvCgK4QQDcAULkvAGSc54B0GJAGUMpCgAoAlEyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5ujABtcNhy9o-aJAA0ylVhXYYiAI6xKnAHIJGyKfPxxEu7AF9nAXWVuXQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare const foo: string | null;
foo && foo.toString();
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/prefer-optional-chain": [
      "error",
      {
        "requireNullish": true
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I believe that this case should be reported. (see _Additional Info_ below)

### Actual Result

This case isn't reported.

### Additional Info

I noticed this while working on https://github.com/typescript-eslint/typescript-eslint/issues/8382: there is an example of incorrect code with `requireNullish: true` - https://typescript-eslint.io/rules/prefer-optional-chain#requirenullish. But it doesn't actually contain any lint errors.

As I understand it, `requireNullish` just restricts the set of types of the left operand of a logical expression to types that are nullish. If we disable `requireNullish`, then the rule reports on the code provided above ([playground](https://typescript-eslint.io/play/#ts=5.3.3&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPJjAJYoDm8APvCgK4QQDcAULkvAGSc54B0GJAGUMpCgAoAlEyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5ujABtcNhy9o-aJAA0ylVhXYYiAI6xKnAHIJGyKfK7jkiXdgC%2BLgLrL3roA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)).

Therefore, I believe that this case should be reported.
